### ISSUE TITLE
TRF: Transformer-based Model with XGBoost for Stock Price Prediction

### ISSUE THREAD

**Issue by ameershaik6667** [Rivier-Computer-Science/AI-Agent-Stock-Prediction#720]
### Overview
Develop a forecasting pipeline that leverages a Transformer-based model to extract temporal features from raw OHLCV (Open, High, Low, Close, Volume) data and uses XGBoost to predict next-day stock price movements. The system will integrate with the existing AI-Agent-Stock-Prediction codebase, provide actionable trading recommendations, and include backtesting using Backtrader to validate performance.

### Description
Traders can input raw OHLCV data for a specific stock ticker, and the system will process it through a Transformer model to learn temporal patterns and dependencies. The Transformer’s output embeddings are fed into an XGBoost model to predict the next-day price movement (up/down) as a classification task. The system outputs a JSON with prediction probabilities, and a CrewAI agent delivers a clear BUY/SELL/HOLD recommendation. Backtesting with Backtrader evaluates the model’s performance over historical data, ensuring actionable and validated trading insights.

### Acceptance Criteria
1.	Data Input: The system ingests raw OHLCV data for a specified ticker, fetched via yfinance, cleaned, normalized, and segmented into fixed-length windows (e.g., 20 days).
2.	Transformer Model: A Transformer-based model (using self-attention) processes the OHLCV windows to generate high-dimensional embeddings capturing temporal patterns.
3.	Embedding Export: The Transformer’s final embeddings are extracted and paired with true labels (e.g., price up/down) for training the XGBoost model.
4.	XGBoost Prediction: The XGBoost classifier uses Transformer embeddings to predict next-day price movement, improving accuracy over the standalone Transformer.
5.	JSON Output: For a given OHLCV window, the system outputs a JSON in the format: 
{ "date": "YYYY-MM-DD", "transformer_score": 0.XX, "xgb_prob": 0.XX }
6.	CrewAI Decision: A CrewAI agent consumes the JSON and outputs a recommendation: 
o	xgb_prob > 0.6 → "BUY"
o	xgb_prob < 0.4 → "SELL"
o	Otherwise → "HOLD"
7.	Backtesting: Backtrader evaluates the model’s predictions over historical data, calculating metrics like accuracy, annualized return, and maximum drawdown.
8.	Visualization: The system generates charts showing predicted vs. actual price movements and backtesting performance (e.g., equity curve).

### Conditions of Satisfaction
1.	Data Processing 
o	Condition: The system fetches, cleans, normalizes, and segments OHLCV data into windows without errors.
o	Test: Fetch OHLCV for a ticker (e.g., AAPL), normalize features (z-score), and create train/validation/test splits with 20-day windows.
o	Satisfaction: Data is correctly processed and usable for model training.
2.	Transformer Model 
o	Condition: The Transformer model trains end-to-end on OHLCV windows, converges on validation data, and produces valid embeddings.
o	Test: Train the model on sample data and verify that validation loss decreases and embeddings have the expected shape.
o	Satisfaction: Model trains without crashing and generates meaningful embeddings.
3.	Embedding Extraction 
o	Condition: The system extracts Transformer embeddings and pairs them with true labels for XGBoost training.
o	Test: Forward test windows through the Transformer and save embeddings to disk (e.g., CSV).
o	Satisfaction: Embeddings are correctly saved and aligned with labels.
4.	XGBoost Training 
o	Condition: XGBoost trains on Transformer embeddings and improves classification accuracy over the standalone Transformer.
o	Test: Compare hold-out accuracy of XGBoost vs. Transformer alone on test data.
o	Satisfaction: XGBoost achieves higher accuracy than the Transformer baseline.
5.	Inference Pipeline 
o	Condition: The inference pipeline processes a new OHLCV window and outputs the required JSON within 5 seconds.
o	Test: Run inference on a sample window and measure latency.
o	Satisfaction: JSON output is correct, and inference completes within 5 seconds.
6.	CrewAI Decision 
o	Condition: The CrewAI agent correctly maps xgb_prob to BUY/SELL/HOLD recommendations.
o	Test: Input sample JSONs with varying xgb_prob values and verify recommendation logic.
o	Satisfaction: Recommendations are consistent with probability thresholds.
7.	Backtesting with Backtrader 
o	Condition: Backtrader evaluates the model’s predictions over historical data, producing metrics like accuracy, return, and drawdown.
o	Test: Run backtesting on a year of historical data and verify metrics align with expected outcomes.
o	Satisfaction: Backtesting results are accurate and include visualizations of performance.
8.	Visualization 
o	Condition: The system generates clear charts showing predicted vs. actual price movements and backtesting performance.
o	Test: Generate charts for a test ticker and verify they accurately reflect predictions and backtesting results.
o	Satisfaction: Charts are interpretable and correctly represent model outputs.
9.	Performance 
o	Condition: The system handles large datasets and multiple predictions without significant latency.
o	Test: Run inference on 100 windows and verify average latency is under 5 seconds.
o	Satisfaction: System performs efficiently under load.
10.	Security 
o	Condition: OHLCV data fetching and storage comply with secure practices.
o	Test: Verify that yfinance API calls use secure connections and data is stored securely.
o	Satisfaction: No vulnerabilities in data handling are detected.

### Definition of Done
1.	Functional Requirements 
o	OHLCV data is fetched, cleaned, normalized, and segmented into windows.
o	Transformer model trains and produces embeddings.
o	XGBoost trains on embeddings and predicts next-day price movements.
o	Inference pipeline outputs JSON with transformer_score and xgb_prob.
o	CrewAI agent delivers BUY/SELL/HOLD recommendations based on xgb_prob.
o	Backtrader backtests predictions and generates performance metrics.
o	Charts visualize predictions and backtesting results.
2.	Non-functional Requirements 
o	Inference completes in ≤ 5 seconds per window.
o	System handles large datasets without performance degradation.
o	Data fetching and storage are secure and compliant with best practices.
3.	Testing and Validation 
o	Unit tests cover data processing, model training, embedding extraction, and inference.
o	Integration tests verify the end-to-end pipeline (data → Transformer → XGBoost → CrewAI → Backtrader).
o	Backtesting results are validated against historical data.
4.	Integration 
o	Code is integrated into the existing AI-Agent-Stock-Prediction codebase on GitHub.
o	Pull request is submitted with all tasks implemented and tested.
5.	Documentation 
o	Documentation covers system architecture, data pipeline, model details, and backtesting setup.
o	Instructions for running the pipeline and interpreting outputs are provided.
6.	Deployment 
o	Pipeline is executable within the existing codebase environment (no Jupyter/Kaggle).
o	All critical bugs are resolved, and no high-priority defects remain.

### Agents Overview
1.	DataAgent
2.	TransformerAgent
3.	EmbeddingAgent
4.	XGBoostAgent
5.	InferenceAgent
6.	DecisionAgent
7.	BacktestAgent
8.	VisualizationAgent

### Detailed Agent Descriptions:
1.	DataAgent 
o	Role: Fetches, cleans, normalizes, and segments OHLCV data into windows.
o	Tasks: 
	Fetch OHLCV data via yfinance.
	Impute missing data and normalize features.
	Create train/validation/test windows.
o	Integration: Uses Python libraries (pandas, yfinance) within the codebase.
o	Interactions: Provides data to TransformerAgent.
2.	TransformerAgent 
o	Role: Trains a Transformer model to extract temporal features from OHLCV windows.
o	Tasks: 
	Implement Transformer architecture with self-attention.
	Train model with regression/classification head.
	Save best model checkpoint.
o	Integration: Uses PyTorch for model implementation.
o	Interactions: Passes embeddings to EmbeddingAgent.
3.	EmbeddingAgent 
o	Role: Extracts and saves Transformer embeddings with corresponding labels.
o	Tasks: 
	Freeze Transformer and forward windows to extract embeddings.
	Save embeddings and labels to disk (CSV).
o	Integration: Uses NumPy/pandas for data handling.
o	Interactions: Provides embeddings to XGBoostAgent.
4.	XGBoostAgent 
o	Role: Trains XGBoost classifier on Transformer embeddings for price movement prediction.
o	Tasks: 
	Load embeddings and labels.
	Perform hyperparameter tuning and train XGBoost.
	Save best model.
o	Integration: Uses XGBoost library with scikit-learn API.
o	Interactions: Provides predictions to InferenceAgent.
5.	InferenceAgent 
o	Role: Processes new OHLCV windows and outputs JSON with predictions.
o	Tasks: 
	Load Transformer and XGBoost models.
	Compute transformer_score and xgb_prob.
	Emit JSON output.
o	Integration: Runs within the codebase environment.
o	Interactions: Passes JSON to DecisionAgent.
6.	DecisionAgent 
o	Role: Maps xgb_prob to BUY/SELL/HOLD recommendations.
o	Tasks: 
	Implement logic: xgb_prob > 0.6 → BUY, < 0.4 → SELL, else HOLD.
	Output JSON with recommendation.
o	Integration: Uses CrewAI for decision logic.
o	Interactions: Receives JSON from InferenceAgent.
7.	BacktestAgent 
o	Role: Runs backtesting using Backtrader to evaluate model performance.
o	Tasks: 
	Implement Backtrader strategy using model predictions.
	Calculate metrics (accuracy, return, drawdown).
o	Integration: Uses Backtrader library.
o	Interactions: Uses predictions from InferenceAgent.
8.	VisualizationAgent 
o	Role: Generates charts for predictions and backtesting results.
o	Tasks: 
	Create charts for predicted vs. actual price movements.
	Generate equity curve and performance metrics visualizations.
o	Integration: Uses matplotlib for plotting.
o	Interactions: Uses data from BacktestAgent and InferenceAgent.

### Agent Interactions and Workflow
1.	DataAgent fetches and preprocesses OHLCV data, passing windows to TransformerAgent.
2.	TransformerAgent trains the model and generates embeddings, sent to EmbeddingAgent.
3.	EmbeddingAgent extracts and saves embeddings, which are used by XGBoostAgent for training.
4.	XGBoostAgent trains the classifier and passes the model to InferenceAgent.
5.	InferenceAgent processes new windows, outputs JSON, and sends it to DecisionAgent.
6.	DecisionAgent generates BUY/SELL/HOLD recommendations.
7.	BacktestAgent evaluates predictions using Backtrader and sends results to VisualizationAgent.
8.	VisualizationAgent creates charts for predictions and backtesting performance.

### Tasks (with Person-Hour Estimates)

- [ ] #721
- [ ] #722
- [ ] #723
- [ ] #724
- [ ] #725
- [ ] #726
- [ ] #727
- [ ] #728
- [ ] #729
- [ ] #730
- [ ] #731
- [ ] #732
- [ ] #733
- [ ] #734
- [ ] #735
- [ ] #736
- [ ] #737
- [ ] #738
- [ ] #739
- [ ] #740
- [ ] #741
- [ ] #742
- [ ] #743
- [ ] #744
- [ ] #745
- [ ] #746
- [ ] #747
- [ ] #748
- [ ] #749
- [ ] #750
- [ ] #751
- [ ] #752
- [ ] #753
- [ ] #754
- [ ] #755
- [ ] #756
- [ ] #757
- [ ] #758
- [ ] #759

Total Estimated Person-Hours: 146 ph

### GitHub Integration
•	All code will be developed as Python scripts compatible with the AI-Agent-Stock-Prediction codebase.
•	Tasks will be submitted as sub-issues under the main issue in the GitHub repository (e.g., #603 for Data Ingestion).
•	A pull request will be created with all implemented tasks, tests, and documentation.

### Notes
•	The Transformer model replaces the CNN-LSTM to address feedback about avoiding duplicated architectures.
•	Backtrader is explicitly included for backtesting, as required.
•	No Jupyter/Kaggle notebooks are used; all development occurs within the codebase.
•	Non-stock use cases and ambiguous terms (e.g., AutoML, optional features) are removed to focus solely on stock price prediction.
•	Charts are generated using matplotlib to visualize predictions and backtesting results, as specified in the VisualizationAgent tasks.
